bin/auto_name_variants.py

```diff
@@ -0,0 +1,185 @@
+    parent = {}
+    papers = collections.defaultdict(list)
+    counts = collections.Counter()
+
+    # The idea is to generate a list of all possible rotations of all
+    # names, e.g., Aravind K. Joshi becomes "aravind k joshi", "k
+    # joshi aravind", "joshi aravind k". Then we sort the list and
+    # compute edit distances between consecutive names on the list,
+    # merging pairs whose distance is below a certain threshold.
+    
+    for filename in sys.argv[1:]:
+        volume = etree.parse(filename).getroot()
+        for paper in volume.findall('paper'):
+            for person in paper.xpath('./author|./editor'):
+                first = text(person.find('first'))
+                last = text(person.find('last'))
+                counts[first, last] += 1
+                papers[first, last].append('{}-{}'.format(volume.attrib['id'], paper.attrib['id']))
+
+                # Make a list of all the normalized name parts
+                full = first + ' ' + last
+                full = normalize(full)
+                full = full.split()
+
+                # Index the name by all possible rotations of the parts
+                for i in range(len(full)):
+                    rotated = ' '.join(full[i:] + full[:i])
+                    keys[rotated].append((first, last))
+
+    for names in keys.values():
+        for name in names:
+            union(parent, name, names[0])
+
+    # Just measure distance between consecutive names
+
+    prevkey = None            
+    for key, names in sorted(keys.items()):
+        if prevkey is not None:
+            d = min(distance(prevkey, key) / len(prevkey),
+                    distance(key, prevkey) / len(key))
+        else:
+            d = float('inf')
+
+        if d < threshold:
+            union(parent, names[0], keys[prevkey][0])
+            print("unify {} = {} (score {})".format(names[0], keys[prevkey][0], d), file=sys.stderr)
+
+        prevkey = key
+
+    clusters = collections.defaultdict(list)
+    for name in parent:
+        clusters[root(parent, name)].append(name)
+
+    # Re-choose canonical names, based on frequency (and then length in case of a tie)
+    
+    newclusters = {}
+    for cluster, names in clusters.items():
+        cname = max(names, key=lambda n: (counts[n], len(n[0])+len(n[1])))
+        newclusters[cname] = names
+    clusters = newclusters
+
+    def paperlist(p):
+        p = sorted(p)
+        if len(p) > 5:
+            return ' '.join(p[:5] + ['...'])
+        else:
+            return ' '.join(p)
+
+    for (cfirst, clast), names in sorted(clusters.items(), key=lambda i: (normalize(i[0][1]), normalize(i[0][0]))):
+        if len(names) > 1 or print_singletons:
+            if print_papers:
+                comment = ' # ' + paperlist(papers[cfirst, clast])
+            print()
+            print('- canonical: {{first: {}, last: {}}}{}'.format(cfirst, clast, comment))
+        if len(names) > 1:
+            print('- variants:')
```

Comment:
That isn't the right output, as it produces a new list entry with just "variants" in it. However, I would *strongly* recommend to do this with a YAML library instead anyway...